src/mesh/coordinates.cxx

```diff
@@ -1619,4 +2017,58 @@ Field2D Coordinates::Laplace_perpXY(const Field2D& A, const Field2D& f) {
   }
 
   return result;
+#else
+  throw BoutException("Coordinates::Laplace_perpXY for 3D metric not implemented");
+#endif
+}
+
+Coordinates::FieldMetric Coordinates::indexDDY(const Field2D& f, CELL_LOC outloc,
+                                                     const std::string& method,
+                                                     const std::string& region) {
+#if BOUT_USE_METRIC_3D
+  if (!f.hasParallelSlices()) {
+    const bool is_unaligned = (f.getDirectionY() == YDirectionType::Standard);
+    const Field3D f_aligned = is_unaligned ? transform->toFieldAligned(f, "RGN_NOX") : f;
+    Field3D result = bout::derivatives::index::DDY(f_aligned, outloc, method, region);
+    return (is_unaligned ? maybeFromFieldAligned(result, region) : result);
+  }
+#endif
+  return bout::derivatives::index::DDY(f, outloc, method, region);
+}
+
+Field3D Coordinates::indexDDY(const Field3D& f, CELL_LOC outloc,
+                              const std::string& method, const std::string& region) {
+#if BOUT_USE_METRIC_3D
+  if (!f.hasParallelSlices()) {
+    const bool is_unaligned = (f.getDirectionY() == YDirectionType::Standard);
+    Field3D f_aligned;
+    if (transform->canToFromFieldAligned()) {
+      f_aligned = is_unaligned ? transform->toFieldAligned(f, "RGN_NOX") : f;
+    } else {
+      Field3D f_parallel = f;
+      transform->calcParallelSlices(f_parallel);
+      return bout::derivatives::index::DDY(f_parallel, outloc, method, region);
+    }
+    Field3D result = bout::derivatives::index::DDY(f_aligned, outloc, method, region);
+    return (is_unaligned ? maybeFromFieldAligned(result, region) : result);
+  }
+#endif
```

Comment:
Just a quick comment as I'm off till the new year -- I added a docstring on this in the header, but basically this is because we need to call `index::DDY` before we've finished constructing the `Coordinates`, so we can't get the parallel transform from it.

Same goes for `maybeFromFieldAligned`.

`Coordinates` has several reimplementations of things because we need to call them in the constructor. One way to get rid of them is to have getters that calculate and cache the results.

Reply:
Another way to get rid of workarounds is to split the creation of `Coordinates` into two parts, so that the derivatives in `Coordinates::geometry()` can use a valid `Coordinates` object. That turned out to be fairly simple - see my proposal in #2179 - because we already have the `Mesh::getCoordinatesSmart()` method that is (I think) the only place the `Coordinates` constructor should be called.